lib/pbench/server/cache_manager.py

```diff
@@ -192,29 +181,155 @@ def make_cache_object(dir_path: Path, path: Path) -> CacheObject:
+        allows lock conversion between LOCK_EX and LOCK_SH.
+
+        Args:
+            lock: the path of a lock file
+        """
+        self.lock = lock.open("w+")
+        self.locked = False
+        self.exclusive = False
+
+    def acquire(self, exclusive: bool = False, wait: bool = True) -> "LockRef":
+        """Acquire the lock
+
+        Args:
+            exclusive: lock for exclusive access
+            wait: wait for lock
+
+        Raises:
+            OSError (EAGAIN or EACCES): wait=False and the lock is already
+                owned
+
+        Returns:
+            The lockref, so this can be chained with the constructor
+        """
+
+        self.exclusive = exclusive
+        cmd = fcntl.LOCK_EX if exclusive else fcntl.LOCK_SH
+        if not wait:
+            cmd |= fcntl.LOCK_NB
+        fcntl.lockf(self.lock, cmd)
+        self.locked = True
+        return self
+
+    def release(self):
+        """Release the lock and close the lock file"""
+
+        if not self.locked:
+            return
+        try:
+            fcntl.lockf(self.lock, fcntl.LOCK_UN)
+            self.lock.close()
+        finally:
+            # Release our reference to the lock file so that the
+            # object can be reclaimed.
+            self.locked = False
+            self.exclusive = False
+            self.lock = None
+
+    def upgrade(self):
+        """Upgrade a shared lock to exclusive"""
+        if not self.exclusive:
+            fcntl.lockf(self.lock, fcntl.LOCK_EX)
+            self.exclusive = True
+
+    def downgrade(self):
+        """Downgrade an exclusive lock to shared"""
+        if self.exclusive:
+            fcntl.lockf(self.lock, fcntl.LOCK_SH)
+            self.exclusive = False
+
+
+class LockManager:
+    def __init__(self, lock: Path, exclusive: bool = False, wait: bool = True):
+        self.lock = LockRef(lock)
+        self.exclusive = exclusive
+        self.wait = wait
+        self.unlock = True
+
+    def keep(self) -> "LockManager":
+        """Tell the context manager not to unlock on exit"""
+        self.unlock = False
+        return self
+
+    def release(self):
+        """Release manually if necessary"""
+        self.lock.release()
+
+    def upgrade(self):
+        """Upgrade a shared lock to exclusive"""
+        self.lock.upgrade()
+        self.exclusive = True
+
+    def downgrade(self):
+        """Downgrade an exclusive lock to shared"""
+        self.lock.downgrade()
+        self.exclusive = False
+
+    def __enter__(self) -> "LockManager":
+        """Enter a lock context manager by acquiring the lock
+
+        Raises:
+            OSError: self.wait is False, and the lock is already owned.
+
+        Returns:
+            the LockManager object
+        """
+        self.lock.acquire(exclusive=self.exclusive, wait=self.wait)
+        return self
+
+    def __exit__(self, *exc):
+        """Exit a lock context manager by releasing the lock
```

Comment:
Apparently this function is [supposed to return a value](https://docs.python.org/3/reference/datamodel.html#object.__exit__)...who knew?  (The default return from a "void" function of `None` turns out to do the right thing for us, but we might want to be more properly explicit, here and in `test_indexing_tarballs.py`.)

Reply:
Actually, this isn't "coincidence" -- from the [PEP](https://peps.python.org/pep-0343/): "The motivation for this detail is to make it possible for mgr.__exit__() to swallow exceptions, without making it too easy (since the default return value, None, is false and this causes the exception to be re-raised)."